src/typing_extensions.py

```diff
@@ -2809,6 +2811,44 @@ def get_protocol_members(tp: type, /) -> typing.FrozenSet[str]:
             return frozenset(tp.__protocol_attrs__)
         return frozenset(_get_protocol_attrs(tp))
 
+if hasattr(typing, "doc"):
+    doc = typing.doc
+    DocInfo = typing.DocInfo
+else:
+    class DocInfo:
+        """Container for documentation information as returned by ``doc()``.
+
+        It is expected this class wouldn't be manually created but instead returned
+        by ``doc()`` inside of type annotations using ``Annotated``.
+
+        The ``documentation`` attribute contains the documentation string passed
+        to ``doc()``.
+        """
+        def __init__(self, documentation: str):
+            self.documentation = documentation
+
+        def __repr__(self) -> str:
+            return f"DocInfo({self.documentation!r})"
+
+    def doc(documentation: str) -> DocInfo:
+        """Define the documentation of a type annotation using ``Annotated``, to be
+         used in class attributes, function and method parameters, return values,
+         and variables.
+
+        The value should be a string literal to allow static tools like editors
+        to use it.
+
+        This complements docstrings.
+
+        It returns a class ``DocInfo`` instance containing the documentation value in
+        the ``documentation`` attribute.
+
+        Example::
+
+            >>> from typing_extensions import doc, Annotated
+            >>> def hi(to: Annotated[str, doc("Who to say hi to")]) -> None: ...
+        """
+        return DocInfo(documentation)
```

Comment:
Just mentioned this on Discuss, but it feels a little redundant to have this function when it doesn't do anything except construct the class. I don't feel too strongly about exactly what color to paint the shed, but obvious alternatives include:

- Make `doc` itself a class, getting rid of `DocInfo` (disadvantage: weird to have a class with a lowercase name)
- Get rid of `doc` and have users call `DocInfo` directly (disadvantage: even more verbose)
- Rename it to `Doc` and have users write `Annotated[param, Doc("my doc")]` (disadvantage: the lowercase `doc` looks nicer to my eye)

I'm sure people on Discuss will come up with other similar ideas. Ultimately it's your call as the PEP author which one of those you prefer.

My main concern here is backward compat for typing-extensions itself: if I merge and release this PR, code that uses `typing_extensions.DocInfo` and `typing_extensions.doc` in a sensible way should keep working essentially in perpetuity. (Because typing-extensions is a low-level dependency for so much of the ecosystem, backward compatibility is very important.) So I want to make sure that whatever we release will keep working even if the PEP's naming scheme changes.

Concretely, the following the operations should keep working, even if we make another typing-extensions release that spells the proposal differently:
- Using `typing_extensions.doc()` to create an instance of the info class
- Using `isinstance(..., typing_extensions.DocInfo)` to check for documentation
- Using `typing_extensions.DocInfo("some doc")` to construct docs (this is not exactly documented but I'm sure people will do it)

I think we can keep these things working with any of my ideas from above, simply by maintaining aliases like `DocInfo = Doc` in a future version of typing-extensions. The one problem is that if only one of `DocInfo` and `doc` ends up in `typing`, then the `if hasattr` above won't work properly, so I think we should check for both separately.

Reply:
FWIW this very much looks to me similar to property in this regard and PEP 8 is fine with classes that look like functions if they are meant to be called like functions.  (I did read the PEP and think it was an odd choice to have a function that does basically nothing)

Reply:
I agree that it seems needlessly duplicative to have a `doc` function and a `DocInfo` class. My personal preference would just be to have a `Doc` class; `Annotated[int, Doc('the number of apples')]` doesn't look too bad to me, and I think we'd quickly get used to it. I'd also be fine if it were a `doc` class, though; I agree with @Gobot1234 that `property`, `classmethod` and `enumerate` are all good precedents here for "function-like" classes that have lowercase names

Reply:
Agreed! And thanks for the feedback regarding this in the discussion as well. I just removed the function and made it a single `Doc` class. I also made it take a positional-only parameter, taking feedback from Eric Traut.

I will update the PEP accordingly later, after finishing with this PR https://github.com/python/peps/pull/3316